File: test/test_listVSdict.py
# test list vs dict
from timeit import default_timer as timer
import array
import numpy as np

class TestListVSDict:
    def __init__(self,size=155):
        self.list = [0]*size
        self.dict = {}
        self.array = array.array('i',[0]*size)
        self.npArray = np.zeros(size,dtype=np.int32)

    def setList(self,i,x):
        self.list[i] = x
    
    def setDict(self,i,x):
        self.dict[i] = x

    def setArray(self,i,x):
        self.array[i] = x

    def setNpArray(self,i,x):
        self.npArray[i] = x


size = 155
repeat = 100000
testListVSDict = TestListVSDict(size)

start = timer()
for rep in range(repeat):
    for i in range(size):
        testListVSDict.setNpArray(i,i)
end = timer()
print("NP Array ",end - start)

start = timer()
for rep in range(repeat):
    for i in range(size):
        testListVSDict.setList(i,i)
end = timer()
print("List ",end - start)


start = timer()
for rep in range(repeat):
    for i in range(size):
        testListVSDict.setDict(i,i)
end = timer()
print("Dict ",end - start)

start = timer()
for rep in range(repeat):
    for i in range(size):
        testListVSDict.setArray(i,i)
end = timer()
print("Array ",end - start)

File: test/test_listVSdict.c
#include <iostream>
#include <vector>
#include <unordered_map>
#include <chrono>
#include <array>

class TestListVSDict {
public:
    TestListVSDict(int size = 155) 
        : list(size, 0), array(), npArray(size, 0) {
        this->size = size;
        for (int i = 0; i < size; ++i) {
            array[i] = 0;
        }
    }

    void setList(int i, int x) {
        list[i] = x;
    }

    void setDict(int i, int x) {
        dict[i] = x;
    }

    void setArray(int i, int x) {
        array[i] = x;
    }

    void setNpArray(int i, int x) {
        npArray[i] = x;
    }

private:
    std::vector<int> list;
    std::unordered_map<int, int> dict;
    std::array<int, 155> array;
    std::vector<int> npArray;
    int size;
};

int main() {
    int size = 155;
    int repeat = 100000;
    TestListVSDict testListVSDict(size);

    auto start = std::chrono::high_resolution_clock::now();
    for (int rep = 0; rep < repeat; ++rep) {
        for (int i = 0; i < size; ++i) {
            testListVSDict.setNpArray(i, i);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    std::cout << "NP Array " << elapsed.count() << std::endl;

    start = std::chrono::high_resolution_clock::now();
    for (int rep = 0; rep < repeat; ++rep) {
        for (int i = 0; i < size; ++i) {
            testListVSDict.setList(i, i);
        }
    }
    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;
    std::cout << "List " << elapsed.count() << std::endl;

    start = std::chrono::high_resolution_clock::now();
    for (int rep = 0; rep < repeat; ++rep) {
        for (int i = 0; i < size; ++i) {
            testListVSDict.setDict(i, i);
        }
    }
    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;
    std::cout << "Dict " << elapsed.count() << std::endl;

    start = std::chrono::high_resolution_clock::now();
    for (int rep = 0; rep < repeat; ++rep) {
        for (int i = 0; i < size; ++i) {
            testListVSDict.setArray(i, i);
        }
    }
    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;
    std::cout << "Array " << elapsed.count() << std::endl;

    return 0;
}
